AMP-129252 [Databricks] Update troubleshooting section to include new topics (#652)
amplitude/amplitude-docs: content/collections/source-catalog/en/databricks.md

   ```
    - **Root cause**: The delta log is truncated and expired, so the Amplitude service didn't find a delta log file to import.
    - **Solution**: Contact Amplitude support to skip expired data and continue the import. By default, retention is 30 days, make sure to keep the data for at least 7 days.

6. ```

>    Fail worker job since databricks job run finished with state MAXIMUM_CONCURRENT_RUNS_REACHED.

Reviewer: let's use raw exception messages we returned to customer, e.g., `Skipping this run because the limit of 1 maximum concurrent runs has been reached`. That makes customer easy to search and find corresponding doc.

Author: @fzqgriffin This is the raw message we returned to customer in the UI.
<img width="1726" alt="Screenshot 2025-04-24 at 10 14 49 AM" src="https://github.com/user-attachments/assets/fe2941ac-2fda-460d-83b9-b343c6acd1cd" />

   ```

>    - **Root cause**: This means your Databricks job is configured to allow only one run at a time. If a new job is triggered while a previous one is still running, it will be skipped. This often occurs when multiple Databricks import sources are set up to pull data simultaneously.

Reviewer: should we call out those those simultaneously jobs are using same cluster as well?

Author: Good point. Will call out this!

   - **Solution**: Update the Databricks job configuration to [increase the maximum concurrent runs](https://docs.databricks.com/aws/en/jobs/configure-job#configure-maximum-concurrent-runs).

7. ```
   [Databricks][JDBCDriver](700100) Connection timeout expired. Details: None.
   ```

>    - **Root cause**: This means Amplitude was unable to establish a JDBC connection to the your databricks endpoint. This often happens because the cluster have been auto-terminated, not started yet or your databricks workspace may have hit limits (e.g., max concurrent SQL endpoints, cluster quota).

Reviewer: > the cluster have been auto-terminated, not started yet

When I investigate last time, cluster will be auto started if it's terminated and usually SQL connection didn't timeout during the time. Can we check if that's true?

Author: I looked into this in more detail. Yes, when a JDBC query is sent to a SQL Warehouse, Databricks will automatically attempt to start the cluster if it’s terminated, and it will queue the request until the cluster is ready.

However, this doesn’t guarantee that the JDBC client will wait long enough. The issue isn’t with Databricks failing to start the cluster. It’s that our JDBC client in Falcon has a connection timeout. If the cluster isn’t ready within that timeout window, the client gives up. So even though Databricks may successfully auto-start the cluster, the connection can still fail if it takes too long.

By default, the connection timeout is 30 seconds. I confirmed that we’re not currently passing it in explicitly: https://github.com/amplitude/nova/blob/a649ef4e99c3ec1754a669ed6a19d57f26863619/src/main/java/com/amplitude/falcon/integration/databricks/DatabricksUtils.java#L153 To handle this better, we can add this `ConnectionTimeout=<time_in_seconds>` to our connection string: https://docs.simba.insightsoftware.com/hc/en-us/articles/30350157836557-Configuring-Connection-Timeouts-and-Retries

Should we just update our connection string to handle this more gracefully, instead of adding it to the troubleshooting? That said, this issue could still occur if Databricks takes longer to start the cluster than the ConnectionTimeout we set. @fzqgriffin Let me know what you think.

Reviewer: @bh202205 bingbing, I agree with you, we should make the change to prevent the error instead of making public doc in that case. Since we are treating JDBC error as something we can mitigate, can we make sure it's not tagged as caused by customer errors? That helps our monitor to catch this so we can proactively mitigate.

Author: @fzqgriffin PR out for adding `ConnectionTimeout` and treat the issue not caused by customer errors: https://github.com/amplitude/nova/pull/22465

>    - **Root cause**: This means Amplitude was unable to establish a JDBC connection to the your databricks endpoint. This often happens because the cluster have been auto-terminated, not started yet or your databricks workspace may have hit limits (e.g., max concurrent SQL endpoints, cluster quota).

Reviewer 2: ```suggestion
   - **Root cause**: This means Amplitude was unable to establish a JDBC connection to the Databricks endpoint. This often happens because the cluster was automatically terminated, not started, or your Databricks workspace may have hit limits (for example, max concurrent SQL endpoints or cluster quota).
```

>    - **Solution**: Check the failed Databricks job run in your workspace. The job details should provide more specific information about the cause of the failure such as issues with cluster availability, query execution, or configuration settings. Once identified, you can address the issue accordingly.

Reviewer: > Check the failed Databricks job run in your workspace
so did it throw JDBC connection error when databricks job run?

Author: @fzqgriffin Ah, I see. Thank you for pointing that out! I realize now that I was confusing the `databricksJobRunId` with the `databricksJobId`. The issue with Zapier occurred because it ran into an error while retrieving Databricks table versions. I’ll go ahead and update the section accordingly to reflect this more accurately.

```
2025-04-23 22:23:46.243 [falcon-worker-6bd5c9445b-q5dcn-processor-daemon-thread-9] ERROR com.amplitude.falcon.worker.inbound.databricks.DatabricksToS3WorkerJobExecutor [{AppId=379088, WorkerJobId=3465223498, WorkerJobType=databricks_to_s3, batchJobId=963278321}] - Error while retrieving Databricks table versions
com.amplitude.falcon.model.exception.databricks.DatabricksJdbcConnectionException: [Databricks][JDBCDriver](700100) Connection timeout expired. Details: None.
```
https://app.datadoghq.com/logs?query=env%3Aprod%20service%3Afalcon-worker%20%223465223498%22&agg_m=count&agg_m_source=base&agg_t=count&cols=host%2Cservice&event=AwAAAZZkwAHmoExZYgAAABhBWlprd0JFeUFBQ2RsSE9uTkx2VEV3RFUAAAAkMDE5NjY0YzMtMjU5My00NGUzLTgzZWYtYWUwNTFmNGEwOWQ1AAfT4A&messageDisplay=inline&refresh_mode=sliding&storage=flex_tier&stream_sort=desc&viz=stream&from_ts=1744910540006&to_ts=1745515340006&live=true

Reviewer: np, @bh202205 , could you please keep our support updated as well so that support team can have right info from us?

>    - **Solution**: Check the failed Databricks job run in your workspace. The job details should provide more specific information about the cause of the failure such as issues with cluster availability, query execution, or configuration settings. Once identified, you can address the issue accordingly.

Reviewer 2: ```suggestion
   - **Solution**: Check the failed Databricks job run in your workspace. The job details should provide more specific information about the cause of the failure such as issues with cluster availability, query execution, or configuration settings.
```

   ```
    - **Root cause**: The delta log is truncated and expired, so the Amplitude service didn't find a delta log file to import.
    - **Solution**: Contact Amplitude support to skip expired data and continue the import. By default, retention is 30 days, make sure to keep the data for at least 7 days.

6. ```
   Fail worker job since databricks job run finished with state MAXIMUM_CONCURRENT_RUNS_REACHED.
   ```

>    - **Root cause**: This means your Databricks job is configured to allow only one run at a time. If a new job run is triggered while a previous run is still in progress, it will be skipped.

Reviewer 2: ```suggestion
   - **Root cause**: Your Databricks job allows only one run at a time. If you trigger a new job while a previous run is still in progress, Databricks skips it. 
```
